includes/REST_API/Link_Controller.php

```diff
@@ -0,0 +1,240 @@
+	 */
+	public function __construct() {
+		$this->namespace = 'web-stories/v1';
+		$this->rest_base = 'link';
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @see register_rest_route()
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'parse_link' ],
+					'permission_callback' => [ $this, 'parse_link_permissions_check' ],
+					'args'                => [
+						'url'       => [
+							'description'       => __( 'The URL to process.', 'web-stories' ),
+							'type'              => 'string',
+							'required'          => true,
+							'sanitize_callback' => 'esc_url_raw',
+						],
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Parses a URL to return some metadata for inserting links.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function parse_link( $request ) {
+		$url = $request['url'];
+
+		$cache_key = 'web_stories_link_data_' . md5( $url );
+		$cache_ttl = DAY_IN_SECONDS;
+		$data      = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			return json_decode( $data, true );
+		}
+
+		$title       = '';
+		$image       = '';
+		$description = '';
+
+		$data = [
+			'title'       => $title,
+			'image'       => $image,
+			'description' => $description,
+		];
+
+		$request = wp_safe_remote_get(
+			$url,
+			[
+				'limit_response_size' => 153600, // 150 KB.
+			]
+		);
+
+		$html  = wp_remote_retrieve_body( $request );
+
+		// Strip <body>.
+		$html_head_end = stripos( $html, '</head>' );
+		if ( $html_head_end ) {
+			$html = substr( $html, 0, $html_head_end );
+		}
+
+		if ( ! $html ) {
+			set_transient( $cache_key, wp_json_encode( $data ), $cache_ttl );
+			return new WP_Error( 'rest_invalid_url', get_status_header_desc( 404 ), array( 'status' => 404 ) );
+		}
+
+
+		$doc                      = new DOMDocument();
+		$doc->strictErrorChecking = false;
+
+		$doc->loadHTML( $html );
+		$xpath = new DOMXpath( $doc );
+
+		// Link title.
+		/* @var DOMNodeList $title_query */
+		$title_query = $xpath->query( '//title' );
+		/* @var DOMNodeList $og_title_query */
+		$og_title_query = $xpath->query( '//meta[@property="og:title"]' );
+		/* @var DOMNodeList $og_site_name_query */
+		$og_site_name_query = $xpath->query( '//meta[@property="og:site_name"]' );
+		if ( $title_query instanceof DOMNodeList && $title_query->length > 0 ) {
+			/** @var \DOMElement $title_node */
+			$title_node = $title_query->item( 0 );
+
+			if ( $title_node instanceof \DOMElement ) {
+				$title = $title_node->textContent;
+			}
+		} else if ( $og_title_query instanceof DOMNodeList && $og_title_query->length > 0 ) {
+			/** @var \DOMElement $title_node */
+			$title_node = $og_title_query->item( 0 );
```

Comment:
A snippet of code the takes a `DOMNodeList` gets the first item and return an attribute is returned 7 times in this PR. Can be make it a private helper method like this.

```
private function get_dom_attribute( $query, $attribute ){
      if ( !( $query instanceof DOMNodeList ) ||  $query->length === 0 ) {
           return false;
      }
       $node = $query->item( 0 );
       if ( $node instanceof \DOMElement ) {
			return $node->getAttribute( $attribute );
	  }
      return false;
}
```

Reply:
Thanks! I was a bit lazy here I suppose 🙂